index.bs

```diff
@@ -2222,6 +2259,84 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
 Issue: Invoke [=parse summary buckets=] and [=parse summary window operator=]
 from this algorithm.
 
+To <dfn>parse max number of event states</dfn> from a [=map=] map,
+a [=source type=] |sourceType|,
+a possibly null 32-bit unsigned integer |attributionScopeLimit|,
+and a [=randomized response output configuration=] |config|:
+1. If |attributionScopeLimit| is null:
+    1. If |map|["`max_num_event_states`"] [=map/exists=], return an error.
+    1. Return null.
+1. Let |states| be the [=obtain a set of possible trigger states|number of possible trigger states=] with |config|.
+1. If |sourceType| is "<code>[=source type/event=]</code>" and |states| is greater than 3 
+    and |map|["`max_num_event_states`"] does not [=map/exist=], return an error.
+1. Let |maxNumEventStates| be the result of running
+    [=parse an optional 32-bit unsigned integer=] with |value|,
+    "`max_num_event_states`", and [=Default max number of event states=].
+1. If |maxNumEventStates| is an error or |maxNumEventStates| is 0 or |maxNumEventStates| is greater than [=Max trigger-state cardinality=], return an error.
+1. If |sourceType| is "<code>[=source type/event=]</code>" and |states| is greater than |maxNumEventStates|, return an error.
+1. Return |maxNumEventStates|.
+
+To <dfn>parse attribution scopes</dfn> from a [=map=] |map| and a possibly null 32-bit unsigned integer |attributionScopeLimit|:
+1. Let |result| be a [=set=].
+1. If |map|["`attribution_scopes`"] [=map/exists=]:
+    1. Let |values| be |map|["`attribution_scopes`"].
+    1. If |values| is not a [=list=], return an error.
+    1. [=list/iterate|For each=] |value| of |values|:
+        1. If |value| is not a string, return an error.
+        1. Let |attributionScope| be the result of applying the
+            <a spec="html">rules for parsing integers</a> to |value|.
+        1. If |attributionScope| cannot be represented by a 64-bit signed integer, return an error.
```

Comment:
Is there a reason we want these scopes to be *signed* integers, instead of unsigned? For comparison, debug keys, trigger data, and `source_event_id` are all unsigned. The only signed 64-bit field we have today is `priority`.

Reply:
Can we ask the same about priority? My sense was that it's similar to priority as both are used for filtering. (It's also easier to support in Java).

Reply:
I believe the most common use cases for attribution scopes are conversion tracking ids, ad group ids and campaign ids. These are unsigned 64-bit afaik. I am leaning towards making this unsigned.

Reply:
The ids we talked about are represented as (positive) 64 *signed* int within Google ads team. I think to be consistent, we should also persist the scopes as 64 *signed* int, and check if they are positive.

(unsigned int64 is slight different from positive signed int64)

Reply:
What would be the reason to insist they are positive?

Reply:
synced offline, and updated this to string to be future proof.